src/main/java/spoon/reflect/visitor/chain/CtQueryImpl.java

```diff
@@ -218,127 +190,184 @@ public int getLength() {
 	 * @param e
 	 * @param parameters
 	 */
-	private void onClassCastException(Step step, ClassCastException e, Object... parameters) {
+	private void onClassCastException(CurrentStep step, ClassCastException e, Object... parameters) {
 		if (step.isFailOnCCE()) {
-			throw e;
+			throw new SpoonException(getStepDescription(step, e.getMessage(), parameters), e);
 		} else if (Launcher.LOGGER.isTraceEnabled()) {
 			//log expected CCE ... there might be some unexpected too!
 			Launcher.LOGGER.trace(e);
 		}
 		log(step, e.getMessage(), parameters);
 	}
 
-	private void log(Step step, String message, Object... parameters) {
+	private void log(CurrentStep step, String message, Object... parameters) {
 		if (isLogging() && Launcher.LOGGER.isInfoEnabled()) {
-			StringBuilder sb = new StringBuilder("Step ");
-			sb.append(step.getName()).append(") ");
-			sb.append(message);
-			for (int i = 0; i < parameters.length; i++) {
-				sb.append("\nParameter ").append(i + 1).append(") ");
-				sb.append(parameters[i]);
+			Launcher.LOGGER.info(getStepDescription(step, message, parameters));
+		}
+	}
+
+	private String getStepDescription(CurrentStep step, String message, Object... parameters) {
+		StringBuilder sb = new StringBuilder("Step ");
+		sb.append(step.getName()).append(") ");
+		sb.append(message);
+		for (int i = 0; i < parameters.length; i++) {
+			sb.append("\nParameter ").append(i + 1).append(") ");
+			if (parameters[i] != null) {
+				sb.append(parameters[i].getClass().getSimpleName());
+				sb.append(": ");
 			}
-			Launcher.LOGGER.info(sb.toString());
+			sb.append(parameters[i]);
 		}
+		return sb.toString();
 	}
 
 	/**
-	 * Each Step must implement it
+	 * The thread local implementation of CtConsumer,
+	 * which knows index of actually processed step
+	 * and handles response of current step and sends it to next step
 	 */
-	private interface Step extends CtConsumer<Object> {
-		/**
-		 * @return next {@link CtConsumer}
-		 */
-		CtConsumer<Object> getNext();
-		/**
-		 * sets next {@link CtConsumer}
-		 */
-		void setNext(CtConsumer<Object> next);
+	private class CurrentStep implements CtConsumer<Object> {

```

Comment:
Ok it took me a while to really understand how the whole process work and I think you should improve the doc on this class as it is central to process the different steps. 
As far as I can see, this class plays as a kind of an orchestrator to move the step cursor forward, get the step, apply it and finally to call the output consumer. It could be great to add those information in the javadoc.